app/src/main/kotlin/at/bitfire/davdroid/sync/Syncer.kt

```diff
@@ -62,86 +66,164 @@ abstract class Syncer {
+     * @return Database collections to be synchronized
+     */
+    abstract fun getSyncCollections(serviceId: Long): List<Collection>
 
-        } finally {
-            if (httpClient.isInitialized())
-                httpClient.value.close()
+    /**
+     * Update an existing local collection with remote collection information
+     *
+     * @param localCollection The local collection to be updated
+     * @param remoteCollection The new remote collection information
+     */
+    abstract fun update(localCollection: CollectionType, remoteCollection: Collection)
 
-            // close content provider client which is acquired above
-            provider.close()
+    /**
+     * Create a new local collection from remote collection information
+     *
+     * @param provider The content provider client to create the local collection
+     * @param remoteCollection The remote collection to be created locally
+     */
+    abstract fun create(provider: ContentProviderClient, remoteCollection: Collection)
 
-            logger.log(
-                Level.INFO,
-                "$contentAuthority sync of $account finished",
-                extras.joinToString(", "))
+    /**
+     * Synchronise local with remote collection contents
+     *
+     * @param provider The content provider client to access the local collection to be updated
+     * @param localCollection The local collection to be synchronized
+     * @param remoteCollection The database collection representing the remote collection. Contains
+     * remote address of the collection to be synchronized.
+     */
+    abstract fun syncCollection(provider: ContentProviderClient, localCollection: CollectionType, remoteCollection: Collection)
+
+    /**
+     * Prepares the sync
+     * - acquires content provider
+     * - handles occurring sync errors
+     */
+    operator fun invoke() {
+        logger.log(Level.INFO, "$authority sync of $account initiated", extras.joinToString(", "))
+
+        // Acquire ContentProviderClient
+        try {
+            context.contentResolver.acquireContentProviderClient(authority)
+        } catch (e: SecurityException) {
+            logger.log(Level.WARNING, "Missing permissions for authority $authority", e)
+            null
+        }.use { provider ->
+            if (provider == null) {
+                /* Can happen if
+                 - we're not allowed to access the content provider, or
+                 - the content provider is not available at all, for instance because the respective
+                   system app, like "calendar storage" is disabled */
+                logger.warning("Couldn't connect to content provider of authority $authority")
+                syncResult.stats.numParseExceptions++ // hard sync error
+
+                return // Don't continue without provider
+            }
+
+            // run sync
+            try {
+                sync(provider)
```

Comment:
Please keep the `/* ose */ if true` because then the code between davx5 and davx5-ose stays (almost) identical.